### ISSUE TITLE
关于DANN(revgrad)中复现代码精度只有55%左右远低于原文中81%中的精度，我对比了一下其他实现找到了一个还原精度的办法。

### ISSUE THREAD

**Issue by Willjay5991** [jindongwang/transferlearning#191]
**修改后的结果**
在修改后的，mnist--》mnist_m 的迁移我训练的时候达到了91%的精度，这个精度是高于原文中的精度的，这是因为我和wang使用的网络结构和结点数目比原文网络不同（比原文网络更复杂）所以获得了更好的精度。

**修改domain classifier 最后输出时的函数为LogSoftmax，而不是这个代码里面的sigmod**
        # domain classifier
        self.domain_classifier = nn.Sequential(
                                        nn.Linear(50*4*4, 100),
                                        nn.BatchNorm1d(100),
                                        nn.ReLU(),                                        
                                        nn.Linear(100,2),
                                        nn.LogSoftmax(dim=1)
                                        )

**修改损失网络的domain classifier, label classifer 的损失函数为NLLLoss，替换wang代码中的CrossEntropyLoss（label classifier）， BCELoss（domain classifier）**

**修改优化器为adam 替换wang的sgd优化器**
我在跑程序的时候，自己对比了一下发现使用RMSprop和Adam优化器可以获得差不多的精度并且收敛速度也快，但是使用SGD就很难收敛。

**Comment by jindongwang** [comment]
很好的结论！感谢！

**Comment by Willjay5991** [comment]
> 
> 
> 很好的结论！感谢！
感谢大佬做的分享，从您知乎专栏的文章中学到了好多。

**Comment by TitanRisen** [comment]
能请教下为什么nn.Linear(100,2)中要做出2维特征的输出吗？ @Willjay5991

**Comment by Willjay5991** [comment]
> 
> 
> 感谢你的建议，试过确实有效。能请教下为什么nn.Linear(100,2)中要做出2维特征的输出吗？或者说为什么原来利用sigmoid函数输出为0-1概率的判别器会不work呢？ @Willjay5991

做出2 dim 输出是用softmax求出数据被判断为 target domain 和 source domain 的概率，概率大的就是最后的分类， 至于为何softmax 比sigmod 好我也没法给出一个理论上的解释。

**Comment by TitanRisen** [comment]
> > 感谢你的建议，试过确实有效。能请教下为什么nn.Linear(100,2)中要做出2维特征的输出吗？或者说为什么原来利用sigmoid函数输出为0-1概率的判别器会不work呢？ @Willjay5991
> 
> 做出2 dim 输出是用softmax求出数据被判断为 target domain 和 source domain 的概率，概率大的就是最后的分类， 至于为何softmax 比sigmod 好我也没法给出一个理论上的解释。

谢谢，我还想请教请问有尝试过最后一层分别求出target domain 和 source domain 的概率在那些情况下会比sigmoid（如，FCN中的pooling+sigmoid）的效果要好？我最近在学习GAN相关的理论，但发现无论是用原始的JS散度还是推土机距离等，判别器的性能还是会有限制。

**Comment by Willjay5991** [comment]
> 
> 
> > > 感谢你的建议，试过确实有效。能请教下为什么nn.Linear(100,2)中要做出2维特征的输出吗？或者说为什么原来利用sigmoid函数输出为0-1概率的判别器会不work呢？ @Willjay5991
> > 
> > 
> > 做出2 dim 输出是用softmax求出数据被判断为 target domain 和 source domain 的概率，概率大的就是最后的分类， 至于为何softmax 比sigmod 好我也没法给出一个理论上的解释。
> 
> 谢谢，我还想请教请问有尝试过最后一层分别求出target domain 和 source domain 的概率在那些情况下会比sigmoid（如，FCN中的pooling+sigmoid）的效果要好？我最近在学习GAN相关的理论，但发现无论是用原始的JS散度还是推土机距离等，判别器的性能还是会有限制。@Willjay5991

没统计过，建议都试试，对于分类问题我个人还是倾向于用softmax，回归问题才选用sigmoid。

**Comment by TitanRisen** [comment]
> > > > 感谢你的建议，试过确实有效。能请教下为什么nn.Linear(100,2)中要做出2维特征的输出吗？或者说为什么原来利用sigmoid函数输出为0-1概率的判别器会不work呢？ @Willjay5991
> > > 
> > > 
> > > 做出2 dim 输出是用softmax求出数据被判断为 target domain 和 source domain 的概率，概率大的就是最后的分类， 至于为何softmax 比sigmod 好我也没法给出一个理论上的解释。
> > 
> > 
> > 谢谢，我还想请教请问有尝试过最后一层分别求出target domain 和 source domain 的概率在那些情况下会比sigmoid（如，FCN中的pooling+sigmoid）的效果要好？我最近在学习GAN相关的理论，但发现无论是用原始的JS散度还是推土机距离等，判别器的性能还是会有限制。@Willjay5991
> 
> 没统计过，建议都试试，对于分类问题我个人还是倾向于用softmax，回归问题才选用sigmoid。

谢谢，另外我试过不用NLLLoss而用CrossEntropy，把discriminator和classifier的最后一层都改为softmax，也能达到一样的效果，精度能达到82%左右，虽然两者的原理应该是一样的。

**Comment by onlyonewater** [comment]
@Willjay5991 ，我这边在复现ADDA的时候，是判别器用softmax不太work，用sigmoid的结果比softmax要好
